onnxscript/rewriter/pattern.py

```diff
@@ -504,13 +572,44 @@ def commute(self) -> list[ValuePattern]:
         return [self]
 
 
+def _nodes_in_pattern(outputs: Sequence[ValuePattern]) -> list[ValuePattern]:
+    """Computes inputs and nodes from outputs of a pattern."""
+    node_patterns: list[NodePattern] = []
+
+    def visit(value_patterns: Sequence[ValuePattern]) -> None:
+        for value_pattern in value_patterns:
+            if isinstance(value_pattern, NodeOutputPattern):
+                node_pattern = value_pattern.producer()
+                if node_pattern not in node_patterns:
+                    node_patterns.append(node_pattern)
+                    visit(node_pattern.inputs)
+
+    visit(outputs)
+    node_patterns.reverse()
+    return node_patterns
+
+
+def _compute_uses(nodepatterns: Sequence[NodePattern]) -> None:
+    """Compute the uses of values in the pattern."""
+    for nodepattern in nodepatterns:
+        for index, value in enumerate(nodepattern.inputs):
+            if value is not None:
+                value.uses().append((nodepattern, index))
+
+
 class GraphPattern:
     """Represents a pattern that can be matched against a subgraph."""
 
-    def __init__(self, outputs: Sequence[ValuePattern]) -> None:
+    def __init__(
+        self, inputs: Sequence[ValuePattern], outputs: Sequence[ValuePattern]
+    ) -> None:
+        self.inputs = inputs
         self.outputs = outputs
         if len(outputs) == 0:
             raise ValueError("GraphPattern must have at least one output")
+        self.nodes = _nodes_in_pattern(outputs)
```

Comment:
Thoughts on making as many attributes private / immutable? I assume the patterns are not to be mutated once created?

Reply:
Yes, I should clean this up!